pkg/controller/cronjob/utils.go

```diff
@@ -110,19 +110,21 @@ func getRecentUnmetScheduleTimes(cj batchv1.CronJob, now time.Time) ([]time.Time
 		// CronJob as last known start time.
 		earliestTime = cj.ObjectMeta.CreationTimestamp.Time
 	}
+
+	latestTime := now
 	if cj.Spec.StartingDeadlineSeconds != nil {
 		// Controller is not going to schedule anything below this point
-		schedulingDeadline := now.Add(-time.Second * time.Duration(*cj.Spec.StartingDeadlineSeconds))
+		schedulingDeadline := sched.Next(earliestTime).Add(time.Second * time.Duration(*cj.Spec.StartingDeadlineSeconds))
 
-		if schedulingDeadline.After(earliestTime) {
-			earliestTime = schedulingDeadline
+		if schedulingDeadline.Before(now) {
+			latestTime = schedulingDeadline
 		}
 	}
-	if earliestTime.After(now) {
+	if earliestTime.After(latestTime) {
 		return []time.Time{}, nil
 	}
 
-	for t := sched.Next(earliestTime); !t.After(now); t = sched.Next(t) {
```

Comment:
I think the code  from https://github.com/kubernetes/kubernetes/blob/v1.20.8/pkg/controller/cronjob/utils.go#L173 to https://github.com/kubernetes/kubernetes/blob/v1.20.8/pkg/controller/cronjob/utils.go#L183 no need to exist here，we  get the all scheduleTimes（if exist） from earliestTime to now is enough，isn't it？

Reply:
There are some differences between controller v1 and v2 versions, and the time has passed a bit long, I am investigating again.